File: src/process-kill.ts
/**
 * ⚠️ LEGACY TypeScript client — frozen. Fix the Rust client instead.
 *
 * The shipping `climon` *client* is the Rust workspace under `rust/` (crates
 * `climon-cli`, `climon-session`, `climon-pty`, `climon-store`, `climon-config`,
 * `climon-remote`, `climon-install`, `climon-update`, …). This module belongs to
 * the legacy Bun/TypeScript client, kept only for local development and the Bun
 * test suite. Do NOT add features or fix client bugs here — make all client
 * changes in the Rust crates. (The Bun dashboard *server* under `src/server*`
 * and `src/web/` is NOT legacy and is still maintained.)
 */
import { spawnSync } from "node:child_process";

export type KillRunner = (cmd: string, args: string[]) => { status: number | null };

const defaultRunner: KillRunner = (cmd, args) => {
  const result = spawnSync(cmd, args, { stdio: "ignore", windowsHide: true });
  return { status: result.error ? 1 : result.status };
};

/**
 * Terminates a process cross-platform, best-effort. On Windows uses `taskkill`
 * with `/T` (whole process tree) and `/F` when forcing. On POSIX sends SIGKILL
 * (force) or SIGTERM (graceful). Returns whether the kill was issued
 * successfully; a process that is already gone reports false on POSIX (ESRCH)
 * and the caller should re-check liveness if it cares.
 *
 * Set `tree: false` on Windows to kill only the specific process without its
 * descendants (e.g. when the caller is itself a child of the target).
 */
export function killProcess(
  pid: number,
  force: boolean,
  platform: NodeJS.Platform = process.platform,
  run: KillRunner = defaultRunner,
  tree: boolean = true
): boolean {
  if (platform === "win32") {
    const args = ["/PID", String(pid)];
    if (tree) args.push("/T");
    if (force) args.push("/F");
    return run("taskkill", args).status === 0;
  }
  try {
    process.kill(pid, force ? "SIGKILL" : "SIGTERM");
    return true;
  } catch {
    return false;
  }
}

/** Returns whether a process with the given pid currently exists. */
export function isProcessAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

File: rust/climon-cli/src/process_kill.rs
//! Cross-platform best-effort process termination. Port of `src/process-kill.ts`.

/// Terminates a process, best-effort. On unix sends `SIGKILL` (force) or
/// `SIGTERM` (graceful) and reports whether the signal was issued without error
/// (a process that is already gone reports `false`/ESRCH). On Windows uses
/// `taskkill /T` (whole tree) plus `/F` when forcing. Mirrors `killProcess`
/// (with `tree: true`).
pub fn kill_process(pid: u32, force: bool) -> bool {
    #[cfg(unix)]
    {
        let signal = if force { libc::SIGKILL } else { libc::SIGTERM };
        unsafe { libc::kill(pid as libc::pid_t, signal) == 0 }
    }
    #[cfg(windows)]
    {
        let mut args: Vec<String> = vec!["/PID".to_string(), pid.to_string(), "/T".to_string()];
        if force {
            args.push("/F".to_string());
        }
        std::process::Command::new("taskkill")
            .args(&args)
            .stdin(std::process::Stdio::null())
            .stdout(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .status()
            .map(|s| s.success())
            .unwrap_or(false)
    }
    #[cfg(not(any(unix, windows)))]
    {
        let _ = (pid, force);
        false
    }
}

/// Returns whether a process with the given pid currently exists. Mirrors
/// `isProcessAlive`.
pub fn is_process_alive(pid: u32) -> bool {
    #[cfg(unix)]
    {
        unsafe { libc::kill(pid as libc::pid_t, 0) == 0 }
    }
    #[cfg(windows)]
    {
        // `tasklist` filtered to the pid prints a header + row when alive and a
        // "no tasks" notice otherwise.
        match std::process::Command::new("tasklist")
            .args(["/FI", &format!("PID eq {pid}"), "/NH"])
            .stdin(std::process::Stdio::null())
            .stderr(std::process::Stdio::null())
            .output()
        {
            Ok(out) => {
                let text = String::from_utf8_lossy(&out.stdout);
                text.contains(&pid.to_string())
            }
            Err(_) => false,
        }
    }
    #[cfg(not(any(unix, windows)))]
    {
        let _ = pid;
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn current_process_is_alive() {
        let pid = std::process::id();
        assert!(is_process_alive(pid));
    }

    #[test]
    fn unlikely_pid_is_not_alive() {
        // A very high pid is almost certainly free.
        assert!(!is_process_alive(4_000_000_000));
    }
}
